File: Java/ds/SparseMatrix.java
import java.util.Scanner;
// Java program for Sparse Matrix Representation using Array 

class SparseMatrix {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System. in );
    System.out.println("Input :");
    System.out.println("Enter the row size");
    int row = sc.nextInt();
    System.out.println("Enter the column size");
    int colmn = sc.nextInt();
    int sparseMatrix[][] = new int[row][colmn];
    System.out.println("Enter the element of 2-D matrix");
    for (int loop = 0; loop < row; loop++) {
      for (int inloop = 0; inloop < colmn; inloop++) {
        sparseMatrix[loop][inloop] = sc.nextInt();
      }
    }
    int count = 0;
    for (int loop = 0; loop < row; loop++) {
      for (int inloop = 0; inloop < colmn; inloop++) {
        if (sparseMatrix[loop][inloop] != 0) {
          count++;
        }
      }
    }
    /* number of columns in Triplet Matrix(count) must be 
         equal to number of non - zero elements in sparseMatrix */
    int tripletMatrix[][] = new int[count][3];
    /* Making of new matrix called TripletMatrix(row,column,value)
          all the non-zero elements are stored in revised Sparse Matrix */
    int index = 0;
    for (int loop = 0; loop < row; loop++) {
      for (int inloop = 0; inloop < colmn; inloop++) {
        if (sparseMatrix[loop][inloop] != 0) {
          tripletMatrix[index][0] = loop;
          tripletMatrix[index][1] = inloop;
          tripletMatrix[index][2] = sparseMatrix[loop][inloop];
          index++;
        }
      }
    }
    System.out.println("Output :");
    System.out.println("Triplet Matrix Representation :-");
    System.out.println("rows columns values");
    for (int loop = 0; loop < count; loop++) {
      for (int inloop = 0; inloop < 3; inloop++) {
        System.out.print(tripletMatrix[loop][inloop]);
      }
      System.out.println();
    }
  }
}
/*
  Input :
  Enter the row size 
  4
  Enter the column size
  5
  Enter the elements of 2-D matrix
  0 0 1 5 0
  0 0 6 0 0
  7 0 0 0 1
  0 8 0 0 0
  Output:
  Triplet Matrix Representation :-
  rows  columns values
  0     2       1
  0     3       5
  1     2       6
  2     0       7
  2     4       1
  3     1       8
 */
/*
  Time Complexity:O(noOfTerms*noOfcolumn)
  Space Complexity:O(1)
 */

File: C/ds/Sparse_Matrix.c
/* SPARSE MATRIX REPRESENTATION USING LINKED LIST
    Representing a sparse matrix by a 2D array leads to wastage of lots of memory.
    A sparse matrix can be effectively represented by storing the non- zero elements
    Here a Linked list is used wherein each node stores
        1. row number
        2. column number
        3. non-zero value
    This representation reduces memory wastage
*/

#include <stdio.h>
#include <stdlib.h>

/* Declare node with row,column and data for storing non-zero term of
   sparse matrix.
   next points to the subsequent node
*/
typedef struct node
{
    int row, column, data;
    struct node *next;
} node;
node *head;

//to represent and print sparse matrix
void sparse_matrix(int rows, int cols, int mat[10][10])
{
    // traverse the matrix row-wise
    for (int i = 0; i < rows; i++)
    {
        for (int j = 0; j < cols; j++)
        {
            //If a non-zero term is encountered
            if (mat[i][j] != 0)
            {
                //Store the non-zero term in a new node of the linked list
                node *new = malloc(sizeof(node));
                new->row = i;
                new->column = j;
                new->data = mat[i][j];
                new->next = NULL;
                //If the list is empty insert as first node
                if (head == NULL)
                    head = n;
                else
                {
                    node *ptr = head;
                    //Traverse till the end of the list
                    while (ptr->next != NULL)
                        ptr = ptr->next;
                    //insert as last node
                    ptr->next = new;
                }
            }
        }
    }
    node *ptr = head;
    printf("\nSPARSE MATRIX REPRESENTATION\nRow\tColumn\tValue\n");
    //Traverse the linked list and print each node
    while (ptr != NULL)
    {
        printf("%d\t%d\t%d\n", ptr->row, ptr->column, ptr->data);
        ptr = ptr->next;
    }
}

//driver code
int main()
{
    node *head = NULL;
    int matrix[10][10], rows, cols, count;
    //accept rows,columns and matrix entries as user input
    scanf("%d%d", &rows, &cols);
    int i, j;
    for (i = 0; i < rows; i++)
    {
        for (j = 0; j < cols; j++)
        {
            scanf("%d", &matrix[i][j]);
        }
    }

    sparse_matrix(rows, cols, matrix);
}

/*
    Sample Input:
    5 5
    0 0 2 0 1
    3 0 0 0 0
    0 0 0 0 0
    0 5 0 4 2
    0 0 6 0 0

    Sample Output:
    Row     Column  Value
    0       2       2
    0       4       1
    1       0       3
    3       1       5
    3       3       4
    3       4       2
    4       2       6
*/

/*
    Time Complexity : O(rows*cols)
    Space Complexity : O(n), where n= number of non-zero terms
*/
